src/Compilers/Core/Portable/MetadataReference/Metadata.cs

```diff
@@ -6,15 +6,34 @@
 namespace Microsoft.CodeAnalysis
 {
     /// <summary>
+    /// An Id that can be used to identify a metadata instance.  If two metadata instances 
+    /// have the same id then they are guaranteed to have the same content.  If two metadata
+    /// instances have different ids, then the contents may or may not be the same.  As such,
+    /// the id is useful as a key in a cache when a client wants to share data for a metadata
+    /// reference as long as it has not changed.
+    /// </summary>
+    public sealed class MetadataId
+    {
```

Comment:
Hey Jared, see the above convo with Matt/Tomas.  My personal main objection over having this be ```object``` is that it gets *super* easy to accidentally use the wrong value as the compiler/language gives you zero type safety.  For example, there is code that looks like this in this pr:  ```base(..., id: reference.Id)```.  If the 'id' were typed as ```object```, it would have been trivial to accidently mistype that as:  ```base(..., id: reference)``` instead.  In that case the compiler would have given me no information that i did anything wrong.  However, because it is typed as MetadataId, i must provide an actual value typed that way.

Similarly, in the code that consumes this later, i have a ```ConditionalWeakTable<MetadataId, ...>```.  In this form the compiler ensures that i'm looking up/storing with actual MetadataIds.  If it were typed as object it would have been trivial to accidently write ```map.GetOrAdd(ref.GetMetadata(), ...)``` instead of ```map.GetOrAdd(ref.GetMetadata().Id, ...)```.

So, in other words, even though it is an empty type, it still provides real type safety in the code.  Something i would like to keep around.  

So personally, i would like it to be typed.  It reduces the chance of errors.  And, if we want to add state to it (perhaps for debugging purposes), now it is trivial to do so.